### PR TITLE
add `make help`

### PR DESCRIPTION
**Pull request by vieux**
Sometimes it's hard to figure out what makefile rule does what.
How about we add a `make test` rule ?

```
$ make help
all                            validate all checks, build linux binaries, run all tests, cross build non-linux binariesand generate archives
binary                         build the linux binaries
cross                          cross build the binaries for darwin, freebsd and windows
win                            cross build the binary for windows
tgz                            build the archives (.zip on windows and .tgz otherwise) containing the binaries
deb                            build the deb packages
docs                           build the docs
gccgo                          build the gcc-go linux binaries
rpm                            build the rpm packages
shell                          start a shell inside the build env
test                           un the unit, integration and docker-py tests
test-docker-py                 run the docker-py tests
test-integration-cli           run the integration tests
test-unit                      run the unit tests
validate                       validate DCO, Seccomp profile generation, gofmt, ./pkg/ isolation, golint, tests, tomls, go vet and vendor
help                           this help
```

ping @tianon

### CODE REVIEW

**Comment by thaJeztah** [comment]
ooooh :heart_eyes:

**Comment by tianon** [comment]
Sweet, LGTM!

I think @jfrazelle will love this too. :+1:

**Comment by jfrazelle** [comment]
YESSSSSS

**Comment by jfrazelle** [comment]
I saw this blog post! <3

**Comment by thaJeztah** [comment]
`make blog` ?

File: Makefile [hunk 1]
```diff
@@ -57,10 +57,10 @@ DOCKER_RUN_DOCKER := $(DOCKER_FLAGS) "$(DOCKER_IMAGE)"
 
 default: binary
 
-all: build
+all: build ## validate all checks, build linux binaries, run all tests, cross build non-linux binariesand generate archives
```
Comment (thaJeztah) [inline Makefile]: can we invent something to make it wrap, so that it fits in smaller windows (80-chars?)

**Comment by thaJeztah** [comment]
tried locally and works, but one note regarding the width of the output :heart:

**Comment by vieux** [comment]
@thaJeztah I added a commit to support \n. It looks like this:

```
$ make help
all                  validate all checks, build linux binaries, run all
                      tests, cross build non-linux binariesand generate archives
binary               build the linux binaries
cross                cross build the binaries for darwin, freebsd and
                      windows
win                  cross build the binary for windows
tgz                  build the archives (.zip on windows and .tgz
                      otherwise) containing the binaries
deb                  build the deb packages
docs                 build the docs
gccgo                build the gcc-go linux binaries
rpm                  build the rpm packages
shell                start a shell inside the build env
test                 un the unit, integration and docker-py tests
test-docker-py       run the docker-py tests
test-integration-cli run the integration tests
test-unit            run the unit tests
validate             validate DCO, Seccomp profile generation, gofmt,
                      ./pkg/ isolation, golint, tests, tomls, go vet and vendor
help                 this help
```

**Comment by duglin** [comment]
I like this!

**Comment by thaJeztah** [comment]
Yup, looks better!

you may need to squash some commits though :wink:

File: Makefile [hunk 2]
```diff
@@ -57,10 +57,10 @@ DOCKER_RUN_DOCKER := $(DOCKER_FLAGS) "$(DOCKER_IMAGE)"
 
 default: binary
 
-all: build
+all: build ## validate all checks, build linux binaries, run all\ntests, cross build non-linux binariesand generate archives
```
Comment (thaJeztah) [inline Makefile]: oh misses a space; "binariesand generate archives"

**Comment by thaJeztah** [comment]
one nit, but LGTM otherwise :+1:

**Comment by jfrazelle** [comment]
lol the blog was
http://marmelab.com/blog/2016/02/29/auto-documented-makefile.html

On Wed, Apr 6, 2016 at 4:56 PM, Sebastiaan van Stijn <
notifications@github.com> wrote:

> one nit, but LGTM otherwise [image: :+1:]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/21821#issuecomment-206625388>
>



-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

**Comment by cpuguy83** [comment]
LGTM

**Pull request merged**